README
What is tdj4?
------------
The letters in "tdj4" stand for "The Daily Journal". It is a PIM program, and
is also the name of the binary executable. It is written using the Qt toolkit from
www.qt.io.

This is version 4-0.1.

More information below, if you still want to read on.

What does PIM mean?
-------------------
Personal Information Manager (I think), a program which manages daily notes,
appointments, alarms, to-do lists, contacts, addresses, etc.

What can it do, in its present form?
------------------------------------
If you are in the habit of keeping daily notes, or a diary, or a journal  of
your activities, tdj can keep them organised and accessible.

That's nice, but does it ever do anything useful?
-------------------------------------------------
If it does, it does it when I'm not looking. Besides, if it did, do you think
I would be giving it away free?

While there is no guarantee that it will do anything useful, or anything at all,
there is every probability of the version number increasing in the future as
and when I get the time to add more features (especially if anyone asks).

How do I use it?
----------------
The user interface is basically a nice, very versatile calendar, with an editor
capable of handling rich text.

You can store your thoughts for a particular day by clicking on the appropriate
date, and entering the data into the editor at the right. To retrieve the notes
for a particular day, click on the desired date. The same editor window displays
the notes, and you can add to or modify the text if needed.

There is a row of tabs which give access to the various components of the PIM.

You can bring up your appointment table for the current day and you can add and
make changes to or delete any appointment.

You can maintain multiple to-do lists, a table of anniversaries as well as names
and addresses of your contacts, sorted into any organized set of categories of
your choice.

Selecting these tabs will display windows where you can enter, modify, or delete
data.

There is a manual also. Read the file "tdjhelp.pdf" in the downloaded package for
more details.

tdj4 is a small and handy program, and takes up very little space on the desktop,
as wellas your hard disk.

Availability
------------
The Daily Journal is available as source code to compile on your own machine, or
as a package containing binaries to install and use.

Compiling the source
--------------------
ALl the source files are available on Github under the following link:
https://github.com/letapk/tdj4

The repository is public. Anyone can check out the branch onto her local machine.

In addition, you will need the libraries and developement files from Qt, which are
available from the depository of your Linux distribution or from www.qt.io.
The current source tree compiles and runs using Qt version 5.15.3.

Also required are the development files and libraries of the GNU Cryptographic library.
They can be found from the GNU website, www.gnu.org, as well as the Linux distribution
depository.

If you are using QtCreator, load the project file tdj4.pro which is available in the source
package. Select the version you want to build (debug version or release version) and hit
Ctrl-B. If there are no problems, this should build the binary.

The program has been compiled and tested on version 5.15.3 of Qt, under Ubuntu
22.04.

The binary may be stored locally in a directoriy separate from the source subdirectory,
depending on how QtCreator is configured. This location will be available within QtCreator.

If you are not using QtCreator, running the comand:
qmake tdj4.pro
should create a Makefile for your system. Then, running:
make
should create a running binary.

When the program is executed, tdj4 searches for a hidden subdirectory called
“.cryptdj”. If this is absent (which will be the case when it is executed the first
time) it issues a notification that it will be created. Subsequently all data is
stored in this subdirectory.

Exit the program and copy the help and image files in the downloaded package to the
hidden data subdirectory. Either use a file manager or open a terminal and do the
following from within the tdj4 subdirectory:

cp tdjhelp.* ~/.cryptdj
cp COPYING ~/.cryptdj

Remember to put the dot (".") in the right place. This will make the on-line help
available when the program is running.

If you are using QtCreator, the location of the locally compiled binary will be
available from within QtCreator. If not, the binary will be in the same location as the
source files. From this location copy the binary to a place from where everyone working
on the computer can run it.

From the subdirectory containing the binary do:
sudo cp ./tdj4 /usr/local/bin/tdj4
Enter the password when prompted.

If your preferred language locale is German, and you wish that the program present
its user interface also in theGerman language, then the file containing the
translations should be in the same location as the program binary, so copy it there
as well:
cp ./tdj4_de.qm /usr/local/bin/tdj4_de.qm

Changes in version 4-0.1
------------------------
1. Added the facility to encrypt user data. This is stored in a subdirectory
different from the one used by version 3-0.3. Data created by each version can
coexist without interference, and since the binaries have different names, they
can also coexist peacefully.

2. Modification of the user manual.

3. One consequence of encryption is that THE FORMAT OF THE DATA FILES IS DIFFERENT.
You cannot read data created by version 3-0.3 into version 4-0.1. However, you can
run both the programs together, side by side, and copy-paste information from one
version to the other.

4. Incorporation of the German language translation in a translation file, which is
used by the user-interface if the language locale is German.

Changes in version 3-0.3
------------------------
1. Added a toolbar and converted the editors from plaintext to rich text. Code for
this has been shamelessly re-used from the Richtext example in the Qt documentation.

A major consequence of the change from plaintext editors to richtext editors is
that the data is stored and interpreted as HTML. This means that earlier notes will
be interpreted as HTML. Since HTML rendering recognizes line breaks and paragraphs
in a different manner, all line breaks and paragraphs in existing notes will be
ignored, and you will have see an unbroken stream of text.

The notes, lists and contact data will have to be re-formatted manually.

3. Added the facility to export the notes and lists as plaintext.

4. Removed the facility to convert data created by the FLTK version "fltdj2" to the
new version 3 format. This is available in version 3-0.2, for those who need to
convert old data.

Changes in version 3-0.2
------------------------
Added compiler directives to detect whether the binary is running under MS-Windows
or Linux, and to set the data path correctly.

Changes in version 3-0.1
------------------------
Converted the user inteface from FLTK to Qt.
This was a complete program rewrite with modifications of the layout in accordance
with the new toolkit.
The notes and appointments data have been split into different files.
Addition of appointments which repeat daily.
The program converts fltdj2 data to the new format once when it is newly installed.
Modification of the user manual.

All versions of The Daily Journal prior to ver. 3 were named "fltdj" and used FLTK,
"The Fast, Light Toolkit" from www.fltk.org for the user interface.

What is the copying policy?
---------------------------
While it is copyright, (C) 2001 onwards, Kartik Patel  the program is also free
software; you can redistribute it and/or modify it under the terms of the GNU
General Public License as published by  the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.  This program
is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.  You
will receive a copy of the GNU General Public License along with this program;
for more information, write to the Free Software  oundation, Inc., 675 Mass
Ave, Cambridge, MA 02139, USA.

What else?
----------
That covers it, more or less. Email me if there is anything left out, or there are
problems in istallation.

Enjoy!

Kartik Patel
letapk@gmail.com
https://letapk.wordpress.com


Changelog.txt
2022
19 May - Rebuild under Qt 5.15.3 on KUbuntu 22.04.
17 May - Commited and pushed some modifications to peripheral files.
16 May - Created repository on Github and uploaded files.
2015
14 May - Modified list display code to use sethtml instead of the earlier conversion to Utf8. This preserves
     user locale and enable umlauts to be displayed. The other editors were already using sethtml.
     Also modified the insert image code to put the absolute path to the html file from the data
     directory .cryptdj, instead of the ":" shortcut, which is not working in the 32-bit version.
2014
20 November - Backport of 4-0.1 richtext to 3-0.3.
20 October - Fixed the notes search code to strip away the HTML markup before looking for the search string.
17 October - The check for zero-length rich text will always fail due to the embedded HTML markup which is not
    displayed in the editor. To check for empty notes or lists, convert to plaintext and then find the length.
14 October - After converting editors to rich text, exporting the notes and lists to text was also adding all
    the HTML markup to the exported files. Fixed by using a QTextDocument to read in the HTML and convert
    to plain text.
13 October - Added font selector and font size controls to toolbar. Added code to correctly indicate on the
    toolbar the format of the text when the cursor position is changed.
10 October - Added a toolbar and converted the three editors from plaintext to rich text.
    set_contact was scrambling the contact tree labels if the contact tab was hidden and a contact
    below a category was directly clicked. Made the tab visible first. Seems to have fixed it.
5 September - Removed all old references to fltk and conversion of FLTK based data. This one is completely Qt.
12 August - Modified the Homepath to the data subdirectory under Linux and Windows. It now contains the
     full path to the data subdirectory.
     It is set once at the beginning of the program and then used everywhere.
     Modified conversion code also to access the new variables.
     A #define is used for TESTING. This accesses a dummy data directory called .tdjtest, instead
     of using the working data directory.
7 August - Added compiler directives to acquire the USERPROFILE environment variable for running under
     Windows, in place of the HOME environment variable under Linux. This is stored in a string called Homepath.
     Also the conversion of data from fltdj2 to tdj3 is now done only under Linux.
4 July - Fixed code to display the daily repeating and non-repeating appointments.
     Added a new class used to sort the rows in the appointments table.
2 July - Added a checkbox to the first column in appointments table. This is to indicate that the
     appointment repeats everyday. Repeating appointments are stored in a separate file.
     Added tooltips to the appointment and anniversary tables.
2013
27 December - Bug in formatting the calendar date which grayed out the background for notes with
     data. Empty notes seem to have a length of 1. Why? Fixed.
26 December - Added a preference and associated code to select the calendar header color.
24 December - Added a preference and associated code to show and remove grid lines in the calendar between dates.
     Added a preference to show or remove the week numbers in the leftmost column.
     Changed some of the key assignments to avoid repeated letters.
20 December - Added code to sort the anniversaries in the order in which they occur. As in the case of the
     To-Do list, any additions or changes while the program runs will be unsorted, but get sorted the next
     time it is executed.
31 October - Added code to sort the To-do list when it is first read from file. Any additions to the list
     while the program runs will be unsorted, but get sorted the next time it is executed.
     Also added a signal to show both the To-do and Contact list after pressing the enter key when it is highlighted.
     This enables moving up and down the list using the cursor keys.
21 August - Debugged alarm related code. Seems to work.
20 August - In the menu Help/Help callback, added searchpath variable that specifies location of resources.
     This helps in display of images within the help file.
     Added functions to issue alarms 5 minutes before appointments.
14 August - Tuned new category and contact messages.
     Added code to trap the Alt-F4 keypress that terminates the application. This will call save_and_quit
     like Alt-File-Exit.
     Ready to roll? Perhaps.
13 August - Debugged the logic of the code which colors the date background. The order matters, and only the first
     call should color the date with the default background, if needed. The later calls overwrite the
     previous ones.
     Adjusted layout of the Search tab.
12 August - Added a button below the calendar displaying today's date and which switches the calendar to today
     when clicked.
     Added code to color dates having notes with a lightgray background.
     Added code to display the currently selected date on the status bar.
8 August - Added a check for the fltdj2 and qtdj2 subdirs at program start. The qtdj2 subdir is created
     if absent. If the fltdj2 subdir is present the program does a conversion of the old data.
     It reads files from the .fltdj2 subdir and writes the data to the .qtdj2 subdirectory in the
     current format. Works. However, the old appointment data (from 2000) onwards is lost, since fltdj2 does
     not read the fltdj1 format. New appointments stored using fltdj2	seem to convert properly.
     The contacts, lists and anniversaries seem to convert without problems.
     Added code to color the days having appointments with a cyan background. For days with both holidays
     and appointments (rare!) the appointment color superceded the holiday color.
6 August - Rearranged prefs tab to create a vbox. This contains the weekstart radiobutton set,

tdjann.cpp
/*

The Daily Journal - A PIM program
Qt version

begin                : 12 July 2013
copyright            : (C) Kartik Patel
email                : letapk@gmail.com

*                                                                         *
*   This program is free software; you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation; either version 2 of the License, or     *
*   (at your option) any later version.                                   *
*                                                                         *

*/

//Last modified 19 June 2022

#include "tdj.h"

extern int get_month_int (QString s);

extern Anniversary anniversary[];

int anniversary_compare (const void *a, const void *b);

void MainWindow::fill_anniversary_items ()
{
int i, days;
QDate *date;

    date = new QDate (calendar->yearShown(), calendar->monthShown(), 1);
    days = date->daysInYear();

    for (i = 0; i < days; i++){
        anncol0[i].setText (anniversary[i+1].month);
        anncol1[i].setText (anniversary[i+1].date);
        anncol2[i].setText (anniversary[i+1].description);
    }
}

void MainWindow::sort_anniversaries (void)
{
    if (max_anniversaries > 0)
        qsort (&(anniversary[1]), max_anniversaries, sizeof (Anniversary), anniversary_compare);
}

int anniversary_compare (const void *a, const void *b)
{
int idate, imonth;
int jdate, jmonth;
Anniversary *c, *d;

    c = (Anniversary *) a;
    d = (Anniversary *) b;

    //find out the integer month in which the holiday falls
    imonth = get_month_int (c->month);
    if (imonth == 0)
        return 0;
    jmonth = get_month_int (d->month);

    if (imonth <  jmonth)
        return -1;
    if (imonth >  jmonth)
        return 1;

    //if the months are the same, check the date
    if (imonth == jmonth) {
        idate = c->date.toInt();
        jdate = d->date.toInt();

        if (idate < jdate)
             return -1;
        if (idate == jdate)
             return 0;
        if (idate > jdate)
            return 1;
    }
    return 0;
}

void MainWindow::get_anniversary_items ()
{
int i = 0, j, row, days;
//QString s;
QDate *date;

    date = new QDate (calendar->yearShown(), calendar->monthShown(), 1);
    days = date->daysInYear();

    for (i = 1; i < 367; i++) {
        anniversary[i].date.clear();
        anniversary[i].month.clear();
        anniversary[i].description.clear();
    }

    //table rows run from 0 onwards
    //but are numbered from 1 onwards on the screen
    j = 1;
    for (row = 0; row < days; row++) {
        //s.clear();
        i = anncol2[row].text().length();

        if (i != 0) {//something to save
            anniversary[j].month = anncol0[row].text();
            anniversary[j].date = anncol1[row].text();
            anniversary[j].description = anncol2[row].text();
            j++;
            i = 0;
        }
    }
}

void MainWindow::save_ann_cell (QTableWidgetItem *it)
{
int row, col;
QString s;

    row = anntable->currentRow();
    col = anntable->currentColumn();

    if (col == -1 || row == -1)
        return;

    s.clear();
    s = it->text();


    if (s.size() != 0) {//something in the cell
        if (col == 0) {
            anncol0[row].setText(s);
            s.clear();
        }
        if (col == 1){
            anncol1[row].setText(s);
            s.clear();
        }
        if (col == 2) {
            anncol2[row].setText(it->text());
            s.clear();
        }
    }
}

ComboBoxItemDelegate::ComboBoxItemDelegate(QObject *parent)
    : QStyledItemDelegate(parent)
{
}

ComboBoxItemDelegate::~ComboBoxItemDelegate()
{
}

QWidget* ComboBoxItemDelegate::createEditor( QWidget *parent, const QStyleOptionViewItem &option, const QModelIndex &index ) const
{
QDate d;

    // Create the combobox
    QComboBox *cb = new QComboBox(parent);

    if(index.column() == 0) {
        //populate it
        cb->addItem(QString (tr("")));
        cb->addItem(QString(tr("January")));
        cb->addItem(QString(tr("February")));
        cb->addItem(QString(tr("March")));
        cb->addItem(QString(tr("April")));
        cb->addItem(QString(tr("May")));
        cb->addItem(QString(tr("June")));
        cb->addItem(QString(tr("July")));
        cb->addItem(QString(tr("August")));
        cb->addItem(QString(tr("September")));
        cb->addItem(QString(tr("October")));
        cb->addItem(QString(tr("November")));
        cb->addItem(QString(tr("December")));
    }

    if(index.column() == 1) {
        cb->addItem(QString(tr("")));

        cb->addItem(QString(tr("1")));
        cb->addItem(QString(tr("2")));
        cb->addItem(QString(tr("3")));
        cb->addItem(QString(tr("4")));
        cb->addItem(QString(tr("5")));
        cb->addItem(QString(tr("6")));
        cb->addItem(QString(tr("7")));

        cb->addItem(QString(tr("8")));
        cb->addItem(QString(tr("9")));
        cb->addItem(QString(tr("10")));
        cb->addItem(QString(tr("11")));
        cb->addItem(QString(tr("12")));
        cb->addItem(QString(tr("13")));
        cb->addItem(QString(tr("14")));

        cb->addItem(QString(tr("15")));
        cb->addItem(QString(tr("16")));
        cb->addItem(QString(tr("17")));
        cb->addItem(QString(tr("18")));
        cb->addItem(QString(tr("19")));
        cb->addItem(QString(tr("20")));
        cb->addItem(QString(tr("21")));

        cb->addItem(QString(tr("22")));
        cb->addItem(QString(tr("23")));
        cb->addItem(QString(tr("24")));
        cb->addItem(QString(tr("25")));
        cb->addItem(QString(tr("26")));
        cb->addItem(QString(tr("27")));
        cb->addItem(QString(tr("28")));

        cb->addItem(QString(tr("29")));
        cb->addItem(QString(tr("30")));
        cb->addItem(QString(tr("31")));
    }

    if(index.column() == 2)
        return QStyledItemDelegate::createEditor(parent, option, index);

    return cb;
}

void ComboBoxItemDelegate::setEditorData ( QWidget *editor, const QModelIndex &index ) const
{
    if(QComboBox *cb = qobject_cast<QComboBox *>(editor)) {
        // get the index of the text in the combobox that matches the current value of the item
        QString currentText = index.data(Qt::EditRole).toString();
        int cbIndex = cb->findText(currentText);
        // if it is valid, adjust the combobox
        if(cbIndex >= 0)
            cb->setCurrentIndex(cbIndex);
    } else {
        QStyledItemDelegate::setEditorData(editor, index);
    }
}

void ComboBoxItemDelegate::setModelData ( QWidget *editor, QAbstractItemModel *model, const QModelIndex &index ) const
{
    if(QComboBox *cb = qobject_cast<QComboBox *>(editor))
        // save the current text of the combo box as the current value of the item
        model->setData(index, cb->currentText(), Qt::EditRole);
    else
        QStyledItemDelegate::setModelData(editor, model, index);
}
